File: langgraph-backend/src/langgraph/email-approval.service.ts
import { Injectable, OnModuleInit } from '@nestjs/common'
import { ChatOpenAI } from '@langchain/openai'
import { ChatOllama } from '@langchain/ollama'
import {
    StateGraph, START, END, Annotation,
    MemorySaver, interrupt, Command,
} from '@langchain/langgraph'
import { HumanMessage } from '@langchain/core/messages'
import { config } from '../config'

const EmailState = Annotation.Root({
    emailRequest: Annotation<string>(),
    draftEmail: Annotation<{ subject: string; recipient: string; body: string }>(),
    approvalStatus: Annotation<'pending' | 'approved' | 'rejected' | 'need_modify'>(),
    modifyFeedback: Annotation<string>(),
    revisionCount: Annotation<number>({
        reducer: (prev, curr) => prev + curr,
        default: () => 0,
    }),
    finalStatus: Annotation<string>(),
})

@Injectable()
export class EmailApprovalService implements OnModuleInit {
    private graph: any

    onModuleInit() {
        // const llm = new ChatOpenAI({
        //   model:         config.langGraph.model,
        //   apiKey:        config.langGraph.apiKey,
        //   configuration: { baseURL: config.langGraph.baseURL },
        //   temperature:   0.7,
        // })
        // 创建 chatOllama 实例
        const llm = new ChatOllama({
            model: config.langGraph.model, // Ollama 模型名称
            temperature: config.langGraph.temperature, // 生成文本的随机程度
            baseUrl: config.langGraph.baseURL, // Ollama 服务器地址
            think: false, // 是否开启思考模式，开启后模型会先返回一个思考中的消息，等生成完成后再返回最终回答
            numPredict: 512, // 生成文本的最大 token 数量，512 是一个比较合理的值，可以根据需要调整
        });

        // ── 节点一：起草邮件 ─────────────────────────────
        // ✅ 节点名改为 draftNode，避免和 State 字段 draftEmail 冲突
        const draftNode = async (state: typeof EmailState.State) => {
            const isRevision = !!state.modifyFeedback
            console.log(`\n✍️  [draftNode] ${isRevision ? '根据修改意见重新起草' : '初次起草'}邮件`)

            const prompt = isRevision
                ? `根据修改意见重新起草邮件：
修改意见：${state.modifyFeedback}
原始需求：${state.emailRequest}
上次草稿：${JSON.stringify(state.draftEmail)}`
                : `根据需求起草一封专业邮件：${state.emailRequest}`

            const res = await llm.invoke([
                new HumanMessage(
                    `${prompt}\n\n输出 JSON（不要其他内容）：
{"subject":"邮件主题","recipient":"收件人","body":"正文内容"}`
                ),
            ])

            let draft: { subject: string; recipient: string; body: string }
            try {
                const json = (res.content as string).replace(/```json\n?|\n?```/g, '').trim()
                draft = JSON.parse(json)
            } catch {
                draft = { subject: '草稿', recipient: '未知', body: res.content as string }
            }

            console.log(`   收件人: ${draft.recipient}，主题: ${draft.subject}`)
            return {
                draftEmail: draft,
                approvalStatus: 'pending' as const,
                revisionCount: isRevision ? 1 : 0,
            }
        }

        // ── 节点二：等待人工审批（interrupt 暂停）──────────
        // ✅ 节点名改为 waitNode，避免和可能的字段名冲突
        const waitNode = async (state: typeof EmailState.State) => {
            console.log(`\n⏸️  [waitNode] 等待人工审批（第 ${state.revisionCount + 1} 版）`)

            const decision = interrupt({
                type: 'email_review',
                message: `请审查邮件草稿（第 ${state.revisionCount + 1} 版）`,
                draft: state.draftEmail,
                options: {
                    approve: '批准发送',
                    reject: '拒绝（取消发送）',
                    modify: '需要修改（附修改意见）',
                },
            })

            console.log(`   人工决定: ${JSON.stringify(decision)}`)

            if (typeof decision === 'string') {
                return { approvalStatus: decision as any }
            }
            if (typeof decision === 'object' && (decision as any)?.action === 'modify') {
                return {
                    approvalStatus: 'need_modify' as const,
                    modifyFeedback: (decision as any).feedback as string,
                }
            }
            return { approvalStatus: 'rejected' as const }
        }

        // ── 路由函数 ──────────────────────────────────────
        const routeAfterApproval = (state: typeof EmailState.State) => {
            console.log(`\n🔀 [route] approvalStatus = ${state.approvalStatus}`)
            switch (state.approvalStatus) {
                case 'approved': return 'sendNode'
                case 'need_modify': return 'draftNode'   // 回到起草节点重新起草
                default: return 'cancelNode'
            }
        }

        // ── 节点三：发送邮件 ──────────────────────────────
        // ✅ 节点名改为 sendNode
        const sendNode = async (state: typeof EmailState.State) => {
            console.log(`\n📤 [sendNode] 发送邮件`)
            console.log(`   收件人: ${state.draftEmail.recipient}`)
            console.log(`   主题:   ${state.draftEmail.subject}`)
            // 实际项目里调用 Nodemailer / SendGrid / 企业邮件 API
            return {
                finalStatus: `✅ 邮件已发送\n收件人：${state.draftEmail.recipient}\n主题：${state.draftEmail.subject}`,
            }
        }

        // ── 节点四：取消发送 ──────────────────────────────
        // ✅ 节点名改为 cancelNode
        const cancelNode = async (state: typeof EmailState.State) => {
            console.log(`\n🚫 [cancelNode] 邮件已取消，状态: ${state.approvalStatus}`)
            return {
                finalStatus: `❌ 邮件已取消（审批状态：${state.approvalStatus}）`,
            }
        }

        this.graph = new StateGraph(EmailState)
            // ✅ 节点名全部改掉，不再和 State 字段名冲突
            .addNode('draftNode', draftNode)
            .addNode('waitNode', waitNode)
            .addNode('sendNode', sendNode)
            .addNode('cancelNode', cancelNode)
            .addEdge(START, 'draftNode')
            .addEdge('draftNode', 'waitNode')
            .addConditionalEdges('waitNode', routeAfterApproval, {
                sendNode: 'sendNode',
                draftNode: 'draftNode',   // 修改意见 → 重新起草（循环）
                cancelNode: 'cancelNode',
            })
            .addEdge('sendNode', END)
            .addEdge('cancelNode', END)
            .compile({ checkpointer: new MemorySaver() })

        console.log('✅ 邮件审批工作流初始化完成')
    }

    // ── 对外方法 ──────────────────────────────────────

    async start(emailRequest: string, threadId: string) {
        console.log(`\n${'═'.repeat(50)}`)
        console.log(`📨 [email/start] threadId: ${threadId}`)
        console.log(`   需求: "${emailRequest}"`)

        const result = await this.graph.invoke(
            { emailRequest },
            { configurable: { thread_id: threadId } }
        )

        if (result.__interrupt__) {
            return {
                status: 'waiting_for_approval',
                threadId,
                reviewData: result.__interrupt__[0].value,
                message: '邮件草稿已生成，请审批',
            }
        }
        return { status: 'completed', result }
    }

    async approve(threadId: string) {
        console.log(`\n✅ [email/approve] threadId: ${threadId}`)
        await this.graph.invoke(
            new Command({ resume: 'approved' }),
            { configurable: { thread_id: threadId } }
        )
        const state = await this.graph.getState({ configurable: { thread_id: threadId } })
        return { status: 'email_sent', finalStatus: state.values.finalStatus }
    }

    async reject(threadId: string) {
        console.log(`\n❌ [email/reject] threadId: ${threadId}`)
        await this.graph.invoke(
            new Command({ resume: 'rejected' }),
            { configurable: { thread_id: threadId } }
        )
        return { status: 'cancelled', message: '邮件已取消发送' }
    }

    async requestModify(threadId: string, feedback: string) {
        console.log(`\n✏️  [email/modify] threadId: ${threadId}`)
        console.log(`   修改意见: "${feedback}"`)
        const result = await this.graph.invoke(
            new Command({ resume: { action: 'modify', feedback } }),
            { configurable: { thread_id: threadId } }
        )
        if (result.__interrupt__) {
            return {
                status: 'waiting_for_approval',
                reviewData: result.__interrupt__[0].value,
                message: '邮件已修改，请重新审批',
            }
        }
        return { status: 'completed' }
    }

    async getState(threadId: string) {
        const state = await this.graph.getState({ configurable: { thread_id: threadId } })
        return state.values
    }
}

File: langgraph-backend-python/app/services/email_approval_service.py
import json
from typing import TypedDict, Literal
from langchain_ollama import ChatOllama
from langgraph.graph import StateGraph, START, END
from langgraph.types import Command, interrupt
from langchain_core.messages import HumanMessage

from app.config import config

llm = ChatOllama(
    model=config["langgraph"]["model"],
    temperature=config["langgraph"]["temperature"],
    base_url=config["langgraph"]["base_url"],
    num_predict=config["langgraph"]["num_predict"],
)

memory_checkpointer = None


class EmailState(TypedDict):
    emailRequest: str
    draftEmail: dict
    approvalStatus: Literal["pending", "approved", "rejected", "need_modify"]
    modifyFeedback: str
    revisionCount: int
    finalStatus: str


_graph_instance = None


def get_graph():
    global memory_checkpointer
    if memory_checkpointer is None:
        from langgraph.checkpoint.memory import MemorySaver
        memory_checkpointer = MemorySaver()

    def draft_node(state: EmailState):
        is_revision = bool(state.get("modifyFeedback"))
        prompt = (
            f"根据修改意见重新起草邮件：\n"
            f"修改意见：{state['modifyFeedback']}\n"
            f"原始需求：{state['emailRequest']}\n"
            f"上次草稿：{json.dumps(state.get('draftEmail', {}))}"
            if is_revision
            else f"根据需求起草一封专业邮件：{state['emailRequest']}"
        )
        res = llm.invoke([
            HumanMessage(
                content=f"{prompt}\n\n输出 JSON（不要其他内容）：\n"
                + '{"subject":"邮件主题","recipient":"收件人","body":"正文内容"}'
            )
        ])
        try:
            clean = res.content.replace("```json", "").replace("```", "").strip()
            draft = json.loads(clean)
        except Exception:
            draft = {"subject": "草稿", "recipient": "未知", "body": res.content}
        return {
            "draftEmail": draft,
            "approvalStatus": "pending",
            "revisionCount": 1 if is_revision else 0,
        }

    def wait_node(state: EmailState):
        decision = interrupt(
            {
                "type": "email_review",
                "message": f"请审查邮件草稿（第 {(state.get('revisionCount') or 0) + 1} 版）",
                "draft": state.get("draftEmail", {}),
                "options": {
                    "approve": "批准发送",
                    "reject": "拒绝（取消发送）",
                    "modify": "需要修改（附修改意见）",
                },
            }
        )
        if isinstance(decision, str):
            return {"approvalStatus": decision}
        if isinstance(decision, dict) and decision.get("action") == "modify":
            return {
                "approvalStatus": "need_modify",
                "modifyFeedback": decision.get("feedback", ""),
            }
        return {"approvalStatus": "rejected"}

    def route_after_approval(state: EmailState):
        if state["approvalStatus"] == "approved":
            return "sendNode"
        if state["approvalStatus"] == "need_modify":
            return "draftNode"
        return "cancelNode"

    def send_node(state: EmailState):
        draft = state.get("draftEmail", {})
        return {
            "finalStatus": (
                f"✅ 邮件已发送\n"
                f"收件人：{draft.get('recipient', '未知')}\n"
                f"主题：{draft.get('subject', '无')}"
            ),
        }

    def cancel_node(state: EmailState):
        return {
            "finalStatus": f"❌ 邮件已取消（审批状态：{state['approvalStatus']}）",
        }

    graph = StateGraph(EmailState)
    graph.add_node("draftNode", draft_node)
    graph.add_node("waitNode", wait_node)
    graph.add_node("sendNode", send_node)
    graph.add_node("cancelNode", cancel_node)
    graph.add_edge(START, "draftNode")
    graph.add_edge("draftNode", "waitNode")
    graph.add_conditional_edges(
        "waitNode",
        route_after_approval,
        {"sendNode": "sendNode", "draftNode": "draftNode", "cancelNode": "cancelNode"},
    )
    graph.add_edge("sendNode", END)
    graph.add_edge("cancelNode", END)
    return graph.compile(checkpointer=memory_checkpointer)


class EmailApprovalService:
    _graph = None

    @classmethod
    def get_graph(cls):
        if cls._graph is None:
            cls._graph = get_graph()
        return cls._graph

    @staticmethod
    async def start(email_request: str, thread_id: str):
        graph = EmailApprovalService.get_graph()
        result = await graph.ainvoke(
            {"emailRequest": email_request},
            config={"configurable": {"thread_id": thread_id}},
        )
        if result.get("__interrupt__"):
            return {
                "status": "waiting_for_approval",
                "threadId": thread_id,
                "reviewData": result["__interrupt__"][0].value,
                "message": "邮件草稿已生成，请审批",
            }
        return {"status": "completed", "result": result}

    @staticmethod
    async def approve(thread_id: str):
        graph = EmailApprovalService.get_graph()
        await graph.ainvoke(
            Command(resume="approved"),
            config={"configurable": {"thread_id": thread_id}},
        )
        state = await graph.aget_state(config={"configurable": {"thread_id": thread_id}})
        return {"status": "email_sent", "finalStatus": state.values.get("finalStatus")}

    @staticmethod
    async def reject(thread_id: str):
        graph = EmailApprovalService.get_graph()
        await graph.ainvoke(
            Command(resume="rejected"),
            config={"configurable": {"thread_id": thread_id}},
        )
        return {"status": "cancelled", "message": "邮件已取消发送"}

    @staticmethod
    async def request_modify(thread_id: str, feedback: str):
        graph = EmailApprovalService.get_graph()
        result = await graph.ainvoke(
            Command(resume={"action": "modify", "feedback": feedback}),
            config={"configurable": {"thread_id": thread_id}},
        )
        if result.get("__interrupt__"):
            return {
                "status": "waiting_for_approval",
                "reviewData": result["__interrupt__"][0].value,
                "message": "邮件已修改，请重新审批",
            }
        return {"status": "completed"}

    @staticmethod
    async def get_state(thread_id: str):
        graph = EmailApprovalService.get_graph()
        state = await graph.aget_state(config={"configurable": {"thread_id": thread_id}})
        return state.values
